Use more node kind sets

Clyybber opened the pull request:
## Summary
* Move  `nkLiterals`  and its more specific literal subsets to 
`ast_types`  and include  `nkNilLit`  in  `nkLiterals` 
* Introduce node kind sets  `nkSIntLiterals` ,  `nkUIntLiterals`  for
signed and unsigned integer literals
* Make more use of node kind set constants across the compiler and make
some case statements exhaustive
* A few minor bugs have been fixed along the way and tests for them
added

## Details
* Fix a bug that made  `astrepr.cyclicTreeAux`  ignore the children of
non-error nodes (only used by the  `astrepr.treeRepr`  debugging
facility)
* Fix the assertion failures encountered when running the  `ic`  tests
with a debug or release compiler
* Fix the  `locks`  pragma erroneously accepting unsigned integers and
char literals (see  `tests/pragmas/tlocks.nim` )
* Fix unsigned integers and char literals causing crashes with term
rewriting macros (see 
`tests/lang_experimental/trmacros/trmacros_various2.nim` )
* Fix the VM crashing when trying to access the non-existent sons of a 
`nnkCommentStmt`  node (see  `tests/errmsgs/tnnodeindexkind.nim` )

Clyybber committed:
ic: Fix assertion failures and include nkCharLit in externUIntLit

Also make case statements exhaustive and define externIntLit in
terms of externSIntLit and externUIntLit

Clyybber committed:
compiler/ast: More exhaustive case statements

Clyybber committed:
sem: More exhaustive case statements

Clyybber committed:
compiler/ast: Make more use of node kind sets

Clyybber committed:
sem: Make more use of node kind sets

Clyybber committed:
compiler: Make more use of node kind sets

Clyybber committed:
Add nkNilLit to nkLiterals

Clyybber committed:
Remove outdated comment

Clyybber committed:
astrepr: Fix cyclicTreeAux ignoring sons for non-nkError nodes

zerbina commented on compiler/ast/trees.nim:
```diff
@@ -49,12 +49,14 @@ proc exprStructuralEquivalent*(a, b: PNode; strictSymEquality=false): bool =
     of nkEmpty, nkNilLit, nkType: result = true
-    else:
+    of nkError:
+      result = true # XXX: This seems wrong
```
Yeah. I'd argue that an error reaching here should be considered a compiler bug.
```suggestion
      unreachable()
```
(`compiler/utils/idioms` needs to be imported).

zerbina commented on compiler/ast/trees.nim:
```diff
@@ -74,16 +76,18 @@ proc sameTree*(a, b: PNode): bool =
     of nkEmpty, nkNilLit, nkType: result = true
-    else:
+    of nkError:
+      result = true # XXX: This seems wrong
```
```suggestion
      unreachable()
```
Same as the suggestion above.

zerbina commented on compiler/ast/trees.nim:
```diff
@@ -99,7 +103,7 @@ proc getMagic*(op: PNode): TMagic =
 
 proc isConstExpr*(n: PNode): bool =
-  const atomKinds = {nkCharLit..nkNilLit} # Char, Int, UInt, Str, Float and Nil literals
+  const atomKinds = nkLiterals # Char, Int, UInt, Str, Float and Nil literals
```
```suggestion
  const atomKinds = nkLiterals
```
The comment is somewhat redundant now, given the declarative set. A very soft suggestion.

zerbina commented on compiler/ic/packed_ast.nim:
```diff
@@ -350,19 +349,11 @@ when false:
-  externSIntLit* = {nkIntLit, nkInt8Lit, nkInt16Lit, nkInt64Lit}
-  externUIntLit* = {nkUIntLit, nkUInt8Lit, nkUInt16Lit, nkUInt32Lit, nkUInt64Lit}
+  externSIntLit* =
+    {nkIntLit, nkInt8Lit, nkInt16Lit, nkInt64Lit} # nkInt32Lit is missing by design!
```
```suggestion
  externSIntLit* = nkSIntLiterals - nkInt32Lit # nkInt32Lit is excluded by design!
```
A soft suggestion. I think it highlights `nkInt32Lit` being excluded a bit better.

zerbina commented on compiler/ic/packed_ast.nim:
```diff
@@ -350,19 +349,11 @@ when false:
+  externSIntLit* =
+    {nkIntLit, nkInt8Lit, nkInt16Lit, nkInt64Lit} # nkInt32Lit is missing by design!
+  externUIntLit* =
+    {nkCharLit, nkUIntLit, nkUInt8Lit, nkUInt16Lit, nkUInt32Lit, nkUInt64Lit}
```
```suggestion
  externUIntLit* = nkUIntLiterals
```

zerbina commented on compiler/sem/guards.nim:
```diff
@@ -451,12 +451,13 @@ proc sameTree*(a, b: PNode): bool =
-    of nkEmpty, nkNilLit: result = true
-    else:
+    of nkError, nkEmpty, nkNilLit, nkCommentStmt:
+      result = true # XXX: Should nkCommentStmt, nkError be handled?
```
The `sameTree` here is meant only for same-ness (in terms of resulting value) of expressions, so comments should be ignored (always 'true') and errors should be treated as unreachable.

zerbina commented on compiler/sem/guards.nim:
```diff
@@ -466,11 +467,11 @@ proc hasSubTree(n, x: PNode): bool =
-    of nkEmpty..nkNilLit:
-      result = n.sameTree(x)
+    of nkWithoutSons - nkFormalParams:
+      result = n.sameTree(x) # XXX: Should nkCommentStmt, nkError be handled?
```
The same as above applies.

zerbina commented on compiler/sem/patterns.nim:
```diff
@@ -62,12 +62,13 @@ proc sameTrees*(a, b: PNode): bool =
-    of nkFloatLit..nkFloat64Lit: result = a.floatVal == b.floatVal
-    of nkStrLit..nkTripleStrLit: result = a.strVal == b.strVal
-    of nkEmpty, nkNilLit: result = true
+    of nkIntLiterals: result = a.intVal == b.intVal
```
Hm, this looks like a bugfix for unsigned integer literals in patterns? (the compiler would have crashed previously, if I'm reading the code correctly)

Clyybber replied:
Yep, added a test now in 03f158a85984ce111a67c8c88c05dabc1b8f9fce

zerbina commented on compiler/sem/patterns.nim:
```diff
@@ -178,12 +179,12 @@ proc matches(c: PPatternContext, p, n: PNode): bool =
-    of nkEmpty, nkNilLit, nkType:
-      result = true
-    else:
+    of nkIntLiterals: result = p.intVal == n.intVal
```
The same comment as above applies.

Clyybber replied:
Yep, added a test now in 03f158a85984ce111a67c8c88c05dabc1b8f9fce

zerbina commented on compiler/sem/patterns.nim:
```diff
@@ -178,12 +179,12 @@ proc matches(c: PPatternContext, p, n: PNode): bool =
+    of nkFloatLiterals: result = p.floatVal == n.floatVal
+    of nkStrLiterals: result = p.strVal == n.strVal
+    of nkError, nkEmpty, nkNilLit, nkType, nkCommentStmt:
+      result = true # XXX: Should nkCommentStmt, nkError be handled?
```
Errors should be treated as unreachable, comments should be ignored in pattern matching, I'd say.

zerbina commented on compiler/sem/sempass2.nim:
```diff
@@ -163,7 +163,7 @@ proc getLockLevel(t: PType): TLockLevel =
   # tyGenericInst(TLock {tyGenericBody}, tyStatic, tyObject):
   if t.kind == tyGenericInst and t.len == 3: t = t[1]
-  if t.kind == tyStatic and t.n != nil and t.n.kind in {nkCharLit..nkInt64Lit}:
+  if t.kind == tyStatic and t.n != nil and t.n.kind in nkIntLiterals:
```
```suggestion
  if t.kind == tyStatic and t.n != nil and t.n.kind in nkSIntLiterals:
```
A minor suggestion. A character literal wouldn't get through `pragmas`.

Clyybber replied:
Due to a bug in `pragmaLock` those could actually pass through. I fixed that bug and changed `getLockLevel` to assert that the locklevel literals are really `nkSIntLiterals` in 5d7c7a1bdeae7dfdd3e86233c7c3bd03cfe12cf1 now.

zerbina commented on compiler/sem/semtempl.nim:
```diff
@@ -954,8 +954,6 @@ proc semTemplBody(c: var TemplCtx, n: PNode): PNode =
 
       if result[i].isError:
         hasError = true
-  of nkError, nkCommentStmt:
-    result = n
```
For anyone wondering, this removal doesn't constitute a change in behaviour (`result` is always set to `n` at the start of the procedure).

zerbina commented on compiler/sem/sighashes.nim:
```diff
@@ -82,22 +82,23 @@ proc hashTree(c: var MD5Context, n: PNode; flags: set[ConsiderFlag]) =
   case n.kind
-  of nkEmpty, nkNilLit, nkType: discard
+  of nkError, nkEmpty, nkNilLit, nkType, nkCommentStmt:
+    discard # XXX: Should nkCommentStmt, nkError be handled?
```
Errors should be treated as impossible (erroneous AST shouldn't be passed to `hashTree`). For comments, I think it'd be better to also treat them as impossible, but the content of a `tyFromExpr` *could* contain them, so it's safer to ignor them for now.

zerbina commented on compiler/sem/sighashes.nim:
```diff
@@ -350,7 +351,8 @@ proc hashBodyTree(graph: ModuleGraph, c: var MD5Context, n: PNode) =
   c &= char(n.kind)
   case n.kind
-  of nkEmpty, nkNilLit, nkType, nkCommentStmt: discard
+  of nkError, nkEmpty, nkNilLit, nkType, nkCommentStmt:
+    discard # XXX: Should nkError be handled?
```
Yep, I think it should be considered impossible.
```suggestion
  of nkError:
    unreachable()
  of nkEmpty, nkNilLit, nkType, nkCommentStmt:
    discard
```

zerbina commented on compiler/utils/astrepr.nim:
```diff
@@ -681,17 +681,17 @@ proc cyclicTreeAux(n: PNode, visited: var seq[PNode], count: var int): bool =
         @[n.diag.wrongNode]
       else:
-        @[]
+        n.sons
```
Perhaps the `sons` local should be a `PNode` here, so that no copy of the sequence is made? (`cyclicTreeAux` is only used for debug rendering, but still)

zerbina commented on compiler/vm/vm.nim:
```diff
@@ -2388,7 +2388,7 @@ proc rawExecute(c: var TCtx, t: var VmThread, pc: var int): YieldReason =
       # TODO: This if-else block should be reordered so as to match the
       #       expectation of occurence
-      if src.kind in {nkEmpty..nkNilLit, nkError}:
+      if src.kind in nkWithoutSons:
```
I think this is a bugfix worth mentioning (maybe a test would make sense too).

Clyybber replied:
Added a test in 1f18c9a975ad78126a9e9b090fc10c1274963750

zerbina commented on doc/astspec.txt:
```diff
@@ -10,7 +10,6 @@ contains:
   type
     NimNodeKind = enum     ## kind of a node; only explanatory
-      nnkNone,             ## invalid node kind
       nnkEmpty,            ## empty node
```
For the sake of accuracy, should `nnkError` also be listed as the first node here?

Clyybber replied:
Thanks, I missed that as my initial approach was to replace nkNone with nkEmpty .
Addressed in https://github.com/nim-works/nimskull/pull/1229/commits/d7c23d9622be02244a018f40d2261fe7091b0dc2 / #1229.

zerbina reviewed (commented):
Very cool. That makes it quite a bit easier to parse the case statements, and future additions of new node kinds become a lot easier. Thank you!

I've answered some questions with what I think would make sense in the asked about situations.

Personally, I think it'd make sense to split this PR into two: one for the set related refactorings (and associated fixes) and one for the removal of `nkNone`.

saem commented on compiler/ast/filters.nim:
```diff
@@ -37,14 +37,16 @@ proc invalidPragma(conf: ConfigRef; n: PNode) =
-    elif i == pos:
-      return n[i]
+  case n.kind
+  of nkWithoutSons: return
```
```suggestion
  of nkWithoutSons: discard "these can't have args"
```

Since it's exhaustive we don't need a `return` (unstructured).

Aside: this proc is a bit weird -- name vs behaviour.

Clyybber committed:
Handle errors in tree equivalence and hash procs

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

Clyybber committed:
Cleanup isConstExpr, externSIntLit, externUIntLit

..and use nkSIntLiterals in getLockLevel

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

Clyybber committed:
Fix pragmaLocks ignoring intLitToIntOrErr errors

..and assert that kind is in nkSIntLiterals in sempass2.getLockLevel
instead of returning 0 for nkUIntLiterals

Clyybber committed:
Expand lock pragma test to cover invalid literals

Clyybber committed:
Elide copy in cyclicTreeAux

Clyybber committed:
Add test for unsigned/char literals in TRM patterns

Clyybber committed:
Fix typo

Clyybber committed:
Add test for indexing error of nnkCommentStmt nodes

zerbina commented on compiler/utils/astrepr.nim:
```diff
@@ -681,19 +681,19 @@ proc cyclicTreeAux(n: PNode, visited: var seq[PNode], count: var int): bool =
+        n
 
-    for nSon in sons:
+    for nSon in nWithSons.sons:
```
```suggestion
    for nSon in nWithSons.items:
```
If the error's `wrongNode` is not a node with sub-nodes, the `sons` access would crash.

saem commented on compiler/ast/ast_types.nim:
```diff
@@ -244,14 +244,18 @@ type
 
 const
+  nkUIntLiterals*  = {nkCharLit, nkUIntLit..nkUInt64Lit}
+  nkSIntLiterals*  = {nkIntLit..nkInt64Lit}
```
```suggestion
  nkSIntLiterals*  = {nkIntLit..nkInt64Lit}
    ## signed int literals
```

I'm a dummy, but it took me far too long to figure out `S` was for signed -- feel free to ignore.

Clyybber committed:
use items iterator in cyclicTreeAux

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

zerbina reviewed (approved):
The changes look good to me. Thanks for splitting the PR up.

For your info, I've fixed a small typo in the PR description (erronously -> erroneously).

saem commented on tests/pragmas/tlocks.nim:
```diff
@@ -11,3 +11,13 @@ method testMethod(g: SomeDerived) =
+  discard
+
+# ensure char literals are rejected
+assert not compiles (;
```
```suggestion
doAssert not compiles (;
```

In case we flip to release build for tests.

saem reviewed (commented):
Few minor comments, splitting a few things out would be a good call.

I really like how the code base reads with these changes, especially removing all the `succ` and `pred` shenanigans.

Clyybber committed:
Remove return and add some comments

Clyybber committed:
Use doAssert in tests/pragmas/tlocks.nim

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

Clyybber commented:
Thanks for the reviews!

Clyybber committed:
Treat nkError reaching patterns.sameTrees as bug

zerbina commented:
@Clyybber: I think also mentioning the introduction of `nkSIntLiterals` and `nkUIntLiterals` in the PR description makes sense, but leaving that up to you.

Apart from that, is this PR good to be merged, @saem?

saem commented:
> Apart from that, is this PR good to be merged, @saem?

Yup, everything else seems good.

Clyybber commented:
> @Clyybber: I think also mentioning the introduction of `nkSIntLiterals` and `nkUIntLiterals` in the PR description makes sense, but leaving that up to you.

Yeah, I added it to the PR description now.

/merge

Clyybber commented:
/merge
